### ISSUE TITLE
Type narrowing based on truthyness can be broken for enum Literals if `__bool__` is overridden

### ISSUE THREAD

**Issue by Daverball** [microsoft/pyright#8074]
**Describe the bug**
It looks like the enum literal special-casing is broken when overriding `__bool__` with a return type of `Literal[False]`

**Code or Screenshots**
Code sample in [pyright playground](https://pyright-play.net/?code=JYWwDg9gTgLgBAUwHYFcQFgBQAzKERwwCeYwSA5nKJLHAIJJEA0cAMsDAlAIYA2WWAMa9uAZ1FwAYhAgAKZGgB0AUVQgAlAC4scXXAAacALxwADAMzCxEgELco8tSrVadewyfOY9cN7oAmCNhwAPohAEYyvGGyogi82OpwALQAfGwcXHwA2pJ8cQC62t4%2BelAIMChQSFL5CBaBwdyyAB6a9IxwAD4ZnDy8uTKK%2BgWuJbrlAG4IfCHEYAit6n5UwW0rPlMz0fOLLcvjiLxxxaVlCNOzu0sNQXDhre0MRN1SMmObF9tzJHsHPsA1qcznAtldfjdDvEThtzpcdhD9rCJhUqjU8sd6pgsI04IJHh0Xj12H0cnYoMNRsCUfCfgtIQCgcjQV9wfSkVDMdTSmCEeyDji7v4Cc9XuSPnDvtcOYy4OtDp9adL-npoQhuYqpYiVWcgA)

```python
import enum
from typing import Any, Literal

class Foo(enum.Enum):
    X = 0

class Bar(enum.Enum):
    X = 0
    
    def __bool__(self) -> Literal[False]:
        return False

def a(x: Any | Literal[Foo.X]):
    reveal_type(x)
    if x:
        reveal_type(x)
    else:
        reveal_type(x)

def b(x: Any | Foo):
    reveal_type(x)
    if x:
        reveal_type(x)
    else:
        reveal_type(x)
        return False

def c(x: Any | Literal[Bar.X]):
    reveal_type(x)
    if x:
        reveal_type(x)
    else:
        reveal_type(x)  # this is wrong, should still be `Any | Literal[Bar.X]`

def d(x: Any | Bar):
    reveal_type(x)
    if x:
        reveal_type(x)
    else:
        reveal_type(x)
```

**Comment by erictraut** [comment]
This is a pretty obscure edge case. It turns out it's relatively easy to fix in pyright, so I'll go ahead and do it. This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.367.
